tsdb/record/record.go

```diff
@@ -184,6 +237,51 @@ func (e *Encoder) Series(series []RefSeries, b []byte) []byte {
 			buf.PutUvarintStr(l.Value)
 		}
 	}
+
+	return buf.Get()
+}
+
+// Metadata encodes series metadata.
+func (e *Encoder) Metadata(series []RefMetadata, b []byte) []byte {
+	buf := encoding.Encbuf{B: b}
+	buf.PutByte(byte(Metadata))
+
+	for _, s := range series {
+		buf.PutUvarint64(s.Ref)
+
+		sizeAfterSeriesRef := buf.Len()
+
+		// Reserve space to write the size of the metadata fields
```

Comment:
This seems a bit over-complicated, and not how we approach this elsewhere. Write the metadata to a temporary buffer, gets its size and then work from there.

Reply:
Are you ok if I make the signature then take two buffers, each that are independent and reused? 

i.e before: 
```go
func (e *Encoder) Metadata(series []RefMetadata, b []byte) []byte {
```

after:
```go
func (e *Encoder) Metadata(series []RefMetadata, b1, b2 []byte) (b1, b2 []byte) {
```

It will be kind of strange however, since callers will only want to use the first result "b1" ("b2" will only be for intermediary buffering on subsequent calls to "Metadata(...)").

The other path I can go down is putting the buffers on the `Encoder` themselves. e.g.
```go
type Encoder struct {
  buffer []byte
  metadataBuffer []byte
}

func NewEncoder(buffer, metadataBuffer []byte) Encoder {
  return Encoder{buffer: buffer, metadataBuffer: metadataBuffer}
}

func (e *Encoder) Bytes() []byte {
  return e.buffer
}

func (e *Encoder) Series(series []RefSeries) {
  // existing implementation but write to e.buffer after clearing, then let caller get the bytes with e.Bytes()
}

func (e *Encoder) Metadata(metadata []RefMetadata) {
  // existing implementation but write to e.buffer after clearing, also use e.metadataBuffer as separate independent buffer which is necessary since we need to write the size of entire record to e.buffer before the contents which is only known after encoding to e.metadataBuffer, then let caller get the bytes with e.Bytes()
}
```